slack-plugin/src/main/kotlin/slack/unittest/UnitTests.kt

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (C) 2022 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package slack.unittest
+
+import com.gradle.enterprise.gradleplugin.testretry.retry as geRetry
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.math.max
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.plugins.AppliedPlugin
+import org.gradle.api.tasks.TaskProvider
+import org.gradle.api.tasks.testing.Test
+import org.gradle.kotlin.dsl.retry
+import org.gradle.language.base.plugins.LifecycleBasePlugin
+import slack.gradle.SlackProperties
+import slack.gradle.ciUnitTestAndroidVariant
+import slack.gradle.isActionsCi
+import slack.gradle.isCi
+import slack.gradle.util.synchronousEnvProperty
+
+/**
+ * This code creates a task named "ciUnitTest" in the project it is applied to, which depends on the
+ * specific unit test task that we want to run on CI for the project.
+ *
+ * Instead of running "./gradlew test" on CI (which would run tests on every build variant on every
+ * project), we prefer to run unit tests on only one build variant to save time.
+ *
+ * For an Android application or library, "ciUnitTest" will depend on "testInternalDebugUnitTest"
+ * (if the "internal"/"external" product flavors exist), or "testDebugUnitTest". For a Java or
+ * Kotlin library, it will depend on "test".
+ *
+ * For convenience, this also creates a task named `compileCiUnitTest` for just _building_ the
+ * tests.
+ *
+ * Finally, this plugin creates a convenience task named "globalCiUnitTest" in the root project,
+ * which depends on all the "ciUnitTest" tasks in the subprojects. This is the task that should be
+ * run on CI.
+ */
+internal object UnitTests {
+  private val MAX_PARALLEL = max(Runtime.getRuntime().availableProcessors() / 2, 1)
```

Comment:
Curious if giving more CPUs to the task is something that we want to experiment with.

Reply:
We could yeah. I'll add an override we could test

Reply:
[b7c2ea9](https://github.com/slackhq/slack-gradle-plugin/pull/280/commits/b7c2ea9fb81666a4b232eb3ae85cce0bea7b1a20)